substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java

```diff
@@ -369,4 +370,48 @@ public MethodEntry ensureMethodEntryForDebugRangeInfo(DebugRangeInfo debugRangeI
     public List<MethodEntry> getMethods() {
         return methods;
     }
+
+    public int lowpc() {
+        assert isPrimary();
+        return primaryEntries.get(0).getPrimary().getLo();
+    }
+
+    public int lowpcDeopt() {
+        assert isPrimary();
+        assert includesDeoptTarget();
+        int lowpc = -1;
+        for (PrimaryEntry primaryEntry : primaryEntries) {
+            Range primary = primaryEntry.getPrimary();
+            if (primary.isDeoptTarget()) {
+                lowpc = primary.getLo();
+                break;
+            }
+        }
+        assert lowpc >= 0;
+        return lowpc;
+    }
+
+    public int hipc() {
+        assert isPrimary();
+        if (!includesDeoptTarget()) {
+            return primaryEntries.get(primaryEntries.size() - 1).getPrimary().getHi();
+        } else {
+            Range lastPrimary = null;
+            for (PrimaryEntry primaryEntry : primaryEntries) {
+                Range primary = primaryEntry.getPrimary();
+                if (primary.isDeoptTarget()) {
+                    break;
+                }
+                lastPrimary = primary;
+            }
+            assert lastPrimary != null;
+            return lastPrimary.getHi();
+        }
+    }
+
+    public int hipcDeopt() {
+        assert isPrimary();
+        assert includesDeoptTarget();
+        return primaryEntries.get(primaryEntries.size() - 1).getPrimary().getHi();
+    }
```

Comment:
Yes, I will add a comment to this effect, although this is already discussed in DebugInfoBase where the choice of a class per compilation unit (as opposed to a file per CU) is explained. I think it is probably best to refer to that comment rather than restate it.

I don't think two lists is a good idea. Classes rarely have large numbers of methods (the exception being a few cases like the reflection holder class where methods are mostly synthetic). Most classes have very few methods and no deopt methods. So adding an extra list per class to avoid a small amount of iteration sounds to me like a bad tradeoff of space against time. I may be wrong but I'd prefer to have some stats to prove me wrong before making any such change.